pkitesting/src/main/java/io/netty/pkitesting/X509Bundle.java

```diff
@@ -0,0 +1,430 @@
+
+    /**
+     * Get the PEM encoded string of the {@linkplain #getCertificate() leaf certificate}.
+     * @return The certificate PEM string.
+     */
+    public String getCertificatePEM() {
+        return toCertPem(certPath[0]);
+    }
+
+    /**
+     * Get the certificate path, starting with the leaf certificate up to but excluding the root certificate.
+     * @return The certificate path.
+     */
+    public X509Certificate[] getCertificatePath() {
+        return certPath.clone();
+    }
+
+    /**
+     * Get the certificate path, starting with the leaf certificate up to and including the root certificate.
+     * @return The certificate path, including the root certificate.
+     */
+    public X509Certificate[] getCertificatePathWithRoot() {
+        X509Certificate[] path = Arrays.copyOf(certPath, certPath.length + 1);
+        path[path.length - 1] = root;
+        return path;
+    }
+
+    /**
+     * Get the certificate path as a list, starting with the leaf certificate up to but excluding the root certificate.
+     * @return The certificate path list.
+     */
+    public List<X509Certificate> getCertificatePathList() {
+        return Arrays.asList(certPath);
+    }
+
+    /**
+     * Get the {@linkplain #getCertificatePath() certificate path} as a PEM encoded string.
+     * @return The PEM encoded certificate path.
+     */
+    public String getCertificatePathPEM() {
+        return toCertPem(certPath);
+    }
+
+    /**
+     * Get the key pair.
+     * @return The key pair.
+     */
+    public KeyPair getKeyPair() {
+        return keyPair;
+    }
+
+    /**
+     * Get the root certificate that anchors the certificate path.
+     * @return The root certificate.
+     */
+    public X509Certificate getRootCertificate() {
+        return root;
+    }
+
+    /**
+     * Get the {@linkplain #getRootCertificate() root certificate} as a PEM encoded string.
+     * @return The PEM encoded root certificate.
+     */
+    public String getRootCertificatePEM() {
+        return toCertPem(root);
+    }
+
+    private static String toCertPem(X509Certificate... certs) {
+        Base64.Encoder encoder = getMimeEncoder();
+        StringBuilder sb = new StringBuilder();
+        for (X509Certificate cert : certs) {
+            sb.append("-----BEGIN CERTIFICATE-----\r\n");
+            try {
+                sb.append(encoder.encodeToString(cert.getEncoded()));
+            } catch (CertificateEncodingException e) {
+                throw new IllegalStateException(e);
+            }
+            sb.append("\r\n-----END CERTIFICATE-----\r\n");
+        }
+        return sb.toString();
+    }
+
+    /**
+     * Get the private key as a PEM encoded PKCS#8 string.
+     * @return The private key in PKCS#8 and PEM encoded string.
+     */
+    public String getPrivateKeyPEM() {
+        Base64.Encoder encoder = getMimeEncoder();
+        StringBuilder sb = new StringBuilder();
+        sb.append("-----BEGIN PRIVATE KEY-----\r\n");
```

Comment:
Let's add an `instance` check here. If the key is EC, make it `BEGIN EC PRIVATE KEY`, and for RSA, make it `BEGIN RSA PRIVATE KEY`.

This will align things with openssl implementation.

Reply:
Yeah, confirmed the EC/RSA bits are only used for PKCS1 keys, and here we only export PKCS8.
See https://www.rfc-editor.org/rfc/rfc7468#section-4